### PR TITLE
nixos/gnunet: use public $TMPDIR

### LINKED ISSUE #133529: GNUnet service doesn't work
### Describe the bug
After enabling the GNUnet service with `services.gnunet.enable = true;`, it prints a lot of errors at startup:
```
Aug 11 17:54:58 nixos gnunet-service-arm[15492]: Aug 11 17:54:58-360075 gns-15492 ERROR Configuration specifies invalid value for option `.pin' in section `gns': Pr>
Aug 11 17:54:58 nixos systemd[1]: Started GNUnet.
Aug 11 17:54:58 nixos gnunet-service-arm[15520]: Aug 11 17:54:58-369385 vpn-15520 ERROR `/nix/store/k8lq1hxqiyf2slnxj6mnw3hkym5vyd0r-gnunet-0.14.1/lib//gnunet/libex>
Aug 11 17:54:58 nixos gnunet-service-arm[15510]: Aug 11 17:54:58-376256 ats-15510 ERROR Assertion failed at plugin_ats_proportional.c:290.
Aug 11 17:54:58 nixos gnunet-service-arm[15510]: Aug 11 17:54:58-376303 ats-15510 ERROR Assertion failed at plugin_ats_proportional.c:290.
Aug 11 17:54:58 nixos gnunet-service-arm[15510]: Aug 11 17:54:58-376332 ats-15510 ERROR Assertion failed at plugin_ats_proportional.c:290.
Aug 11 17:54:58 nixos gnunet-service-arm[15510]: Aug 11 17:54:58-376358 ats-15510 ERROR Assertion failed at plugin_ats_proportional.c:290.
Aug 11 17:54:58 nixos gnunet-service-arm[15517]: Aug 11 17:54:58-376330 transport-tcp-15517 WARNING Unexpected address length: 24 bytes
Aug 11 17:54:58 nixos gnunet-service-arm[15517]: Aug 11 17:54:58-376366 transport-15517 ERROR Assertion failed at gnunet-service-transport_validation.c:902.
Aug 11 17:54:58 nixos gnunet-service-arm[15510]: Aug 11 17:54:58-376380 ats-15510 ERROR Assertion failed at plugin_ats_proportional.c:290.
Aug 11 17:54:58 nixos gnunet-service-arm[15517]: Aug 11 17:54:58-376383 transport-15517 ERROR Address with 24 bytes for plugin tcp and peer DSTJ is malformed
Aug 11 17:54:58 nixos gnunet-service-arm[15517]: Aug 11 17:54:58-376390 transport-tcp-15517 WARNING Unexpected address length: 12 bytes
Aug 11 17:54:58 nixos gnunet-service-arm[15517]: Aug 11 17:54:58-376404 transport-15517 ERROR Assertion failed at gnunet-service-transport_validation.c:902.
Aug 11 17:54:58 nixos gnunet-service-arm[15510]: Aug 11 17:54:58-376405 ats-15510 ERROR Assertion failed at plugin_ats_proportional.c:290.
Aug 11 17:54:58 nixos gnunet-service-arm[15517]: Aug 11 17:54:58-376409 transport-15517 ERROR Address with 12 bytes for plugin tcp and peer DSTJ is malformed
Aug 11 17:54:58 nixos gnunet-service-arm[15510]: Aug 11 17:54:58-376428 ats-15510 ERROR Assertion failed at plugin_ats_proportional.c:290.
Aug 11 17:54:58 nixos gnunet-service-arm[15510]: Aug 11 17:54:58-376451 ats-15510 ERROR Assertion failed at plugin_ats_proportional.c:290.
```
Additionally, attempting to stop the deamon with `gnunet-arm -e` reveals that it is apparently not even running: 
```
Aug 11 20:21:58-538831 gnunet-arm-19435 WARNING GNUnet not running, cannot stop the peer
```

### Metadata
Please run `nix-shell -p nix-info --run "nix-info -m"` and paste the result.

```
 - system: `"x86_64-linux"`
 - host os: `Linux 5.10.42, NixOS, 21.11.20210610.432fc2d (Porcupine)`
 - multi-user?: `yes`
 - sandbox: `yes`
 - version: `nix-env (Nix) 2.4pre20210601_5985b8b`
 - channels(root): `"nixos-21.05pre279456.04a2b269d89, release-20.09.tar.gz-20.09"`
 - nixpkgs: `/nix/var/nix/profiles/per-user/root/channels/nixos`
```

### PR DESCRIPTION
**Pull request by malte-v**
###### Motivation for this change
The `gnunet-*` CLI utilities connect to the daemon via sockets in `$TMPDIR` (usually `/tmp`). If we use a private systemd-supplied `$TMPDIR`, the CLI utilities can't connect to the daemon and thus don't work correctly.

Resolves #133529

###### Things done

- [x] Tested using sandboxing ([nix.useSandbox](https://nixos.org/nixos/manual/options.html#opt-nix.useSandbox) on NixOS, or option `sandbox` in [`nix.conf`](https://nixos.org/nix/manual/#sec-conf-file) on non-NixOS linux)
- Built on platform(s)
   - [x] NixOS
   - [ ] macOS
   - [ ] other Linux distributions
- [ ] Tested via one or more NixOS test(s) if existing and applicable for the change (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review wip"`
- [ ] Tested execution of all binary files (usually in `./result/bin/`)
- [21.11 Release Notes (or backporting 21.05 Release notes)](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md#generating-2111-release-notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

### CODE REVIEW

**Comment by malte-v** [comment]
After merging this, the CLI utilities can be used as follows: `sudo -u gnunet sh -c 'TMPDIR=/tmp gnunet-arm -I'`

**Comment by rnhmjoj** [comment]
Can't the sockets be moved under a more appropriate directory?
The typical place for such files would be `/run/gnunet`. You can easily manage this in systemd by setting `RuntimeDirectory=gnunet`.

**Comment by malte-v** [comment]
@rnhmjoj Done.

**Comment by rnhmjoj** [comment]
I'd like to suggest a few more changes:
1. There's a `GNUNET_USER_RUNTIME_DIR` variable to change the socket directory specifically
2. Linking the configuration to /etc/gnunet.conf makes sure all cli tools read the right directory
3. Use `StateDirectory=gnunet` instead of `createHome`
4. Use `/var/lib/gnunet/data`, as suggested in the gnunet [documentation](https://docs.gnunet.org/handbook/gnunet.html#Peer-configuration-for-distributors-_0028e_002eg_002e-Operating-Systems_0029) (this probably needs to be mentioned in the release notes)

Here's a patch:
```patch
diff --git a/nixos/modules/services/networking/gnunet.nix b/nixos/modules/services/networking/gnunet.nix
index b154537e524..ea68ffece72 100644
--- a/nixos/modules/services/networking/gnunet.nix
+++ b/nixos/modules/services/networking/gnunet.nix
@@ -6,12 +6,15 @@ let
 
   cfg = config.services.gnunet;
 
-  homeDir = "/var/lib/gnunet";
+  stateDir = "/var/lib/gnunet";
 
-  configFile = with cfg; pkgs.writeText "gnunetd.conf"
+  configFile = with cfg;
     ''
       [PATHS]
-      SERVICEHOME = ${homeDir}
+      GNUNET_HOME = ${stateDir}
+      GNUNET_RUNTIME_DIR = /run/gnunet
+      GNUNET_USER_RUNTIME_DIR = /run/gnunet
+      GNUNET_DATA_HOME = ${stateDir}/data
 
       [ats]
       WAN_QUOTA_IN = ${toString load.maxNetDownBandwidth} b
@@ -137,8 +140,6 @@ in
     users.users.gnunet = {
       group = "gnunet";
       description = "GNUnet User";
-      home = homeDir;
-      createHome = true;
       uid = config.ids.uids.gnunet;
     };
 
@@ -148,17 +149,19 @@ in
     # so install them globally.
     environment.systemPackages = [ cfg.package ];
 
+    environment.etc."gnunet.conf".text = configFile;
+
     systemd.services.gnunet = {
       description = "GNUnet";
       after = [ "network.target" ];
       wantedBy = [ "multi-user.target" ];
       path = [ cfg.package pkgs.miniupnpc ];
-      environment.TMPDIR = "/run/gnunet";
-      serviceConfig.ExecStart = "${cfg.package}/lib/gnunet/libexec/gnunet-service-arm -c ${configFile}";
+      serviceConfig.ExecStart = "${cfg.package}/lib/gnunet/libexec/gnunet-service-arm -c /etc/gnunet.conf";
       serviceConfig.User = "gnunet";
       serviceConfig.UMask = "0007";
-      serviceConfig.WorkingDirectory = homeDir;
+      serviceConfig.WorkingDirectory = stateDir;
       serviceConfig.RuntimeDirectory = "gnunet";
+      serviceConfig.StateDirectory = "gnunet";
     };
 
   };

```

File: nixos/modules/services/networking/gnunet.nix [hunk 1]
```diff
@@ -148,17 +149,19 @@ in
     # so install them globally.
     environment.systemPackages = [ cfg.package ];
 
+    environment.etc."gnunet.conf".text = configFile;
+
     systemd.services.gnunet = {
       description = "GNUnet";
       after = [ "network.target" ];
       wantedBy = [ "multi-user.target" ];
       path = [ cfg.package pkgs.miniupnpc ];
```
Comment (aanderse) [inline nixos/modules/services/networking/gnunet.nix:158]:
```suggestion
      wantedBy = [ "multi-user.target" ];
      restartTriggers = [ configFile ];
      path = [ cfg.package pkgs.miniupnpc ];
```

The code in this PR is inadvertently tricky. Changes to your configuration will not impact any running services anymore. Just pop in a `restartTriggers` and you're good to go!

File: nixos/modules/services/networking/gnunet.nix [hunk 2]
```diff
@@ -6,12 +6,15 @@ let
 
   cfg = config.services.gnunet;
 
-  homeDir = "/var/lib/gnunet";
+  stateDir = "/var/lib/gnunet";
 
-  configFile = with cfg; pkgs.writeText "gnunetd.conf"
+  configFile = with cfg;
```
Comment (aanderse) [inline nixos/modules/services/networking/gnunet.nix:11]: This module is just *begging* for a `settings` option! Any takers?

**Review by aanderse — CHANGES_REQUESTED** [review]

**Comment by Artturin** [comment]
Is this ready? What @aanderse said can be done in a separate pr

**Comment by malte-v** [comment]
Yeah I think it's ready, works fine for me so far

**Pull request merged**